// @webex/internal-plugin-presence/dist/constants.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
var GROUNDSKEEPER_INTERVAL = exports.GROUNDSKEEPER_INTERVAL = 20 * 1000; // 20 seconds in ms

var FETCH_DELAY = exports.FETCH_DELAY = 300; // ms

var UPDATE_PRESENCE_DELAY = exports.UPDATE_PRESENCE_DELAY = 60 * 1000; // 1 minute in ms
var EXPIRED_PRESENCE_TIME = exports.EXPIRED_PRESENCE_TIME = 10 * 60 * 1000; // 10 minutes in ms

var SUBSCRIPTION_DELAY = exports.SUBSCRIPTION_DELAY = 60 * 1000; // 1 minute in ms
var PREMATURE_EXPIRATION_SUBSCRIPTION_TIME = exports.PREMATURE_EXPIRATION_SUBSCRIPTION_TIME = 60 * 1000; // 1 minute in ms
var DEFAULT_SUBSCRIPTION_TTL = exports.DEFAULT_SUBSCRIPTION_TTL = 10 * 60 * 1000; // 10 minutes in ms

var APHELEIA_SUBSCRIPTION_UPDATE = exports.APHELEIA_SUBSCRIPTION_UPDATE = 'event:apheleia.subscription_update';
var PRESENCE_UPDATE = exports.PRESENCE_UPDATE = 'updated';

var ENVELOPE_TYPE = exports.ENVELOPE_TYPE = {
  SUBSCRIPTION: 'subscription',
  PRESENCE: 'presence',
  DELETE: 'delete'
};
