// @webex/internal-plugin-presence/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _has2 = require('lodash/has');

var _has3 = _interopRequireDefault(_has2);

var _presence = require('./presence');

Object.defineProperty(exports, 'default', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_presence).default;
  }
});

require('@webex/internal-plugin-device');

var _webexCore = require('@webex/webex-core');

var _presence2 = _interopRequireDefault(_presence);

var _config = require('./config');

var _config2 = _interopRequireDefault(_config);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/*!
 * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
 */

(0, _webexCore.registerInternalPlugin)('presence', _presence2.default, {
  payloadTransformer: {
    predicates: [{
      name: 'normalizeSingleStatusResponse',
      direction: 'inbound',
      test: function test(ctx, response) {
        // POST to /apheleia/api/v1/events
        return _promise2.default.resolve((0, _has3.default)(response, 'body.eventType') && (0, _has3.default)(response, 'body.subject'));
      },
      extract: function extract(response) {
        return _promise2.default.resolve(response);
      }
    }],
    transforms: [{
      name: 'normalizeSingleStatusResponse',
      direction: 'inbound',
      fn: function fn(ctx, response) {
        response.body.status = response.body.eventType;
      }
    }]
  },
  config: _config2.default
});
