// @webex/internal-plugin-device/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FeaturesModel = exports.FeatureModel = exports.FeatureCollection = exports.DeviceUrlInterceptor = exports.constants = exports.config = exports.default = undefined;

var _device = require('./device');

Object.defineProperty(exports, 'default', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_device).default;
  }
});

var _webexCore = require('@webex/webex-core');

var _device2 = _interopRequireDefault(_device);

var _index = require('./features/index');

var _deviceUrl = require('./interceptors/device-url');

var _deviceUrl2 = _interopRequireDefault(_deviceUrl);

var _constants = require('./constants');

var constants = _interopRequireWildcard(_constants);

var _config = require('./config');

var _config2 = _interopRequireDefault(_config);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// Internal dependencies.
(0, _webexCore.registerInternalPlugin)('device', _device2.default, {
  config: _config2.default,
  interceptors: {
    DeviceUrlInterceptor: _deviceUrl2.default.create
  },
  /**
   * Unregister the device in the case that the webex instance has logged out.
   *
   * @returns {Promise<undefined>}
   */
  onBeforeLogout: function onBeforeLogout() {
    return this.unregister();
  }
});

// Plugin dependencies.
exports.config = _config2.default;
exports.constants = constants;
exports.DeviceUrlInterceptor = _deviceUrl2.default;
exports.FeatureCollection = _index.FeatureCollection;
exports.FeatureModel = _index.FeatureModel;
exports.FeaturesModel = _index.FeaturesModel;
