// @webex/internal-plugin-device/dist/features/features-model.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

var _ampersandState = require('ampersand-state');

var _ampersandState2 = _interopRequireDefault(_ampersandState);

var _constants = require('../constants');

var _featureCollection = require('./feature-collection');

var _featureCollection2 = _interopRequireDefault(_featureCollection);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * Feature collection parent container.
 *
 * @description
 * This class contains all of the feature collection class objects to help
 * organize the data retrieved from the **wdm** service on device registration.
 */


// Local Dependencies
var FeaturesModel = _ampersandState2.default.extend({

  // Ampersand property members.

  collections: {
    /**
     * This collection contains the developer feature collection.
     *
     * @type {FeatureCollection}
     */
    developer: _featureCollection2.default,

    /**
     * This collection contains the entitlement feature collection.
     *
     * @type {FeatureCollection}
     */
    entitlement: _featureCollection2.default,

    /**
     * This collection contains the user feature collection.
     *
     * @type {FeatureCollection}
     */
    user: _featureCollection2.default
  },

  // Helper method members.

  /**
   * Recursively clear attributes, children, and collections.
   *
   * @param {Object} options - Attribute options to unset.
   * @returns {this}
   */
  clear: function clear(options) {
    var _this = this;

    // Clear the ampersand attributes safely if there are any. This class should
    // never have any attributes.
    (0, _keys2.default)(this.attributes).forEach(function (key) {
      _this.unset(key, options);
    });

    // Clear the ampersand children safely if there are any. This class should
    // never have any children.
    /* eslint-disable-next-line no-underscore-dangle */
    (0, _keys2.default)(this._children).forEach(function (key) {
      _this[key].clear();
    });

    // Clear the ampersand collections safely.
    /* eslint-disable-next-line no-underscore-dangle */
    (0, _keys2.default)(this._collections).forEach(function (key) {
      _this[key].reset();
    });

    return this;
  },


  // Ampersand method members.

  /**
   * Initializer method for FeatureModel class object.
   *
   * @override
   * @returns {void}
   */
  initialize: function initialize() {
    var _this2 = this;

    // Declare the collection event names.
    var eventNames = ['change:value', 'add', 'remove'];

    // Initialize collection event listeners.
    eventNames.forEach(function (eventName) {
      _constants.FEATURE_COLLECTION_NAMES.forEach(function (collectionName) {
        _this2[collectionName].on(eventName, function (model, options) {
          _this2.trigger('change:' + collectionName, _this2, _this2[collectionName], options);
        });
      });
    });
  }
}); // External dependencies.
exports.default = FeaturesModel;
