// @webex/plugin-meetings/dist/locus-info/selfUtils.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _constants = require('../constants');

var _parameter = require('../common/errors/parameter');

var _parameter2 = _interopRequireDefault(_parameter);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var SelfUtils = {};

/**
 * parses the relevant values for self: muted, guest, moderator, mediaStatus, state, joinedWith, creator, id
 * @param {Object} self
 * @param {String} deviceId
 * @returns {undefined}
 */
SelfUtils.parse = function (self, deviceId) {
  if (self) {
    return {
      muted: SelfUtils.getMuted(self),
      lastModified: SelfUtils.getLastModified(self),
      modifiedBy: SelfUtils.getModifiedBy(self),
      guest: self.guest,
      moderator: self.moderator,
      mediaStatus: SelfUtils.getStatus(self.status),
      state: self.state,
      // TODO: give a proper name . With same device as login or different login`
      // Some times we might have joined with both mobile and web
      joinedWith: self.devices.find(function (device) {
        return deviceId === device.url;
      }),
      creator: self.isCreator, // check if its used,
      selfId: self.id,
      selfIdentity: SelfUtils.getSelfIdentity(self),
      selfUrl: self.url,
      removed: self.removed
    };
  }

  return null;
};

SelfUtils.getSelves = function (oldSelf, newSelf, deviceId) {
  var previous = oldSelf && SelfUtils.parse(oldSelf, deviceId);
  var current = newSelf && SelfUtils.parse(newSelf, deviceId);
  var updates = {};

  updates.isUnadmittedGuest = SelfUtils.isUnadmittedGuest(current);
  updates.isAdmittedGuest = SelfUtils.isAdmittedGuest(previous, current);
  updates.isMutedByOthers = SelfUtils.mutedByOthers(previous, current);
  updates.moderatorChanged = SelfUtils.moderatorChanged(previous, current);

  return {
    previous: previous,
    current: current,
    updates: {
      isUnadmittedGuest: SelfUtils.isUnadmittedGuest(current),
      isAdmittedGuest: SelfUtils.isAdmittedGuest(previous, current),
      mutedByOthers: SelfUtils.mutedByOthers(previous, current)
    }
  };
};

SelfUtils.getLastModified = function (self) {
  if (!self || !self.controls || !self.controls.audio || !self.controls.audio.meta || !self.controls.audio.meta.lastModified) {
    return null;
  }

  return self.controls.audio.meta.lastModified;
};

SelfUtils.getModifiedBy = function (self) {
  if (!self || !self.controls || !self.controls.audio || !self.controls.audio.meta || !self.controls.audio.meta.modifiedBy) {
    return null;
  }

  return self.controls.audio.meta.modifiedBy;
};

/**
 * get the id from the self object
 * @param {Object} self
 * @returns {String}
 */
SelfUtils.getSelfIdentity = function (self) {
  if (!self && !self.person) {
    return null;
  }

  return self.person.id;
};

/**
 * get the muted property from the self object
 * @param {Object} self
 * @returns {Boolean}
 */
SelfUtils.getMuted = function (self) {
  if (!self || !self.controls || !self.controls.audio) {
    return null;
  }

  return self.controls.audio.muted;
};

SelfUtils.getStatus = function (status) {
  return {
    audio: status.audioStatus,
    video: status.videoStatus,
    slides: status.videoSlidesStatus
  };
};

/**
 * @param {Object} check
 * @returns {Boolean}
 */
SelfUtils.isLocusGuestUnadmitted = function (check) {
  return check && check.guest && check.state === _constants._IDLE_;
};

/**
 * @param {Object} check
 * @returns {Boolean}
 */
SelfUtils.isLocusGuestAdmitted = function (check) {
  return check && check.guest && check.state === _constants._JOINED_;
};

/**
 * @param {Object} self
 * @returns {Boolean}
 * @throws {Error} when self is undefined
 */
SelfUtils.isUnadmittedGuest = function (self) {
  if (!self) {
    throw new _parameter2.default('self must be defined to determine if self is unadmitted as guest.');
  }

  return SelfUtils.isLocusGuestUnadmitted(self);
};

SelfUtils.moderatorChanged = function (oldSelf, changedSelf) {
  if (!oldSelf) {
    return true;
  }
  if (!changedSelf) {
    throw new _parameter2.default('New self must be defined to determine if self transitioned moderator status.');
  }

  return oldSelf.moderator !== changedSelf.moderator;
};

/**
 * @param {Object} oldSelf
 * @param {Object} changedSelf
 * @returns {Boolean}
 * @throws {Error} if changed self was undefined
 */
SelfUtils.isAdmittedGuest = function (oldSelf, changedSelf) {
  if (!oldSelf) {
    // if there was no previous locus, it couldn't have been admitted yet
    return false;
  }
  if (!changedSelf) {
    throw new _parameter2.default('New self must be defined to determine if self transitioned to admitted as guest.');
  }

  return SelfUtils.isLocusGuestUnadmitted(oldSelf) && SelfUtils.isLocusGuestAdmitted(changedSelf);
};

SelfUtils.mutedByOthers = function () {
  var oldSelf = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var changedSelf = arguments[1];

  if (!changedSelf) {
    throw new _parameter2.default('New self must be defined to determine if self was muted by others.');
  }

  return changedSelf.muted && (oldSelf.lastModified !== changedSelf.lastModified || oldSelf.modifiedBy !== changedSelf.modifiedBy);
};

/**
 * extract the sipUrl from the partner
 * @param {Object} partner
 * @param {Object} info
 * @returns {Object}
 */

SelfUtils.getSipUrl = function (partner, type, sipUri) {
  // For webex meeting the sipUrl gets updated in info parser
  if (partner && type === _constants._CALL_) {
    return { sipUri: partner.person.sipUrl || partner.person.emailAddress || partner.person.email };
  }

  return { sipUri: sipUri };
};

exports.default = SelfUtils;
