// @webex/webex-core/dist/interceptors/auth.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _promise = require('babel-runtime/core-js/promise');

var _promise2 = _interopRequireDefault(_promise);

var _deleteProperty = require('babel-runtime/core-js/reflect/delete-property');

var _deleteProperty2 = _interopRequireDefault(_deleteProperty);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _httpCore = require('@webex/http-core');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @class
 */
var AuthInterceptor = function (_Interceptor) {
  (0, _inherits3.default)(AuthInterceptor, _Interceptor);

  function AuthInterceptor() {
    (0, _classCallCheck3.default)(this, AuthInterceptor);
    return (0, _possibleConstructorReturn3.default)(this, (AuthInterceptor.__proto__ || (0, _getPrototypeOf2.default)(AuthInterceptor)).apply(this, arguments));
  }

  (0, _createClass3.default)(AuthInterceptor, [{
    key: 'onRequest',


    /**
     * @see {@link Interceptor#onRequest}
     * @param {Object} options
     * @returns {Object}
     */
    value: function onRequest(options) {
      var _this2 = this;

      options.headers = options.headers || {};

      // If Authorizations is already set, don't overwrite it
      if ('authorization' in options.headers || 'auth' in options) {
        // If Authorization is set to null, false, or undefined, delete it to
        // prevent a CORS preflight.
        if (!options.headers.authorization) {
          (0, _deleteProperty2.default)(options.headers, 'authorization');
        }

        return _promise2.default.resolve(options);
      }

      return this.requiresCredentials(options).then(function (requires) {
        if (!requires) {
          return options;
        }

        return _this2.webex.credentials.getUserToken().then(function (token) {
          options.headers.authorization = token.toString();

          return options;
        });
      });
    }

    /**
     * Determines if the provided options object needs an authorization header.
     *
     * @param {Object} options
     * @returns {Promise<boolean>}
     */

  }, {
    key: 'requiresCredentials',
    value: function requiresCredentials(options) {
      var _this3 = this;

      // Validate that authorization is necessary.
      if (options.addAuthHeader === false) {
        return _promise2.default.resolve(false);
      }

      // Validate that the services plugin has been loaded before proceeding.
      if (!this.webex.internal.services) {
        return _promise2.default.resolve(false);
      }

      // Destructure webex instance to isolate services plugin.
      var services = this.webex.internal.services;

      // Store the current service details if available and destructure details.

      var details = services.getServiceFromUrl(options.uri || '');

      var _ref = details || {},
          name = _ref.name;

      var resource = options.resource,
          uri = options.uri;

      var service = options.service || options.api;

      if (service && service === 'web-authentication' || name && name === 'web-authentication') {
        return _promise2.default.resolve(false);
      }

      // Unique validation for the u2c service.
      if (service && service === 'u2c' || name && name === 'u2c') {
        if (resource && resource.includes('limited') || uri && uri.includes('limited')) {
          return _promise2.default.resolve(false);
        }

        return _promise2.default.resolve(true);
      }

      // Validate that the allowed domains can be utilized.
      if (services.validateDomains && services.hasAllowedDomains() && uri && services.isAllowedDomainUrl(uri)) {
        return _promise2.default.resolve(true);
      }

      // Perform an additional validation in case the service does not exist yet.
      return services.waitForService({ name: service, url: uri }).then(function (detectedUrl) {
        // Validate if the url was detected while waiting for the service.
        if (detectedUrl) {
          return true;
        }

        // Return false to indicate authentication is not required.
        return false;
      }).catch(function (error) {
        _this3.webex.logger.warn('auth-interceptor: failed to validate service exists in catalog', error);

        return false;
      });
    }

    /**
     * @see {@link Interceptor#onResponseError}
     * @param {Object} options
     * @param {Error} reason
     * @returns {Object}
     */

  }, {
    key: 'onResponseError',
    value: function onResponseError(options, reason) {
      var _this4 = this;

      return this.shouldAttemptReauth(reason, options).then(function (shouldAttemptReauth) {
        if (shouldAttemptReauth) {
          _this4.webex.logger.info('auth: received 401, attempting to reauthenticate');

          if (reason.options.headers) {
            (0, _deleteProperty2.default)(reason.options.headers, 'authorization');
          }

          if (_this4.webex.credentials.canRefresh) {
            return _this4.webex.credentials.refresh().then(function () {
              return _this4.replay(options);
            });
          }
        }

        return _promise2.default.reject(reason);
      });
    }

    /**
     * Replays the request
     * @param {Object} options
     * @returns {Object}
     */

  }, {
    key: 'replay',
    value: function replay(options) {
      if (options.replayCount) {
        options.replayCount += 1;
      } else {
        options.replayCount = 1;
      }

      if (options.replayCount > this.webex.config.maxAuthenticationReplays) {
        this.webex.logger.error('auth: failed after ' + this.webex.config.maxAuthenticationReplays + ' replay attempts');

        return _promise2.default.reject(new Error('Failed after ' + this.webex.config.maxAuthenticationReplays + ' replay attempts'));
      }

      this.webex.logger.info('auth: replaying request ' + options.replayCount + ' time');

      return this.webex.request(options);
    }

    /**
     * Indicates whether or not the current request should refresh its access
     * token in event of a 401
     * @param {Error} reason
     * @param {Object} options
     * @returns {Promise<boolean>}
     */

  }, {
    key: 'shouldAttemptReauth',
    value: function shouldAttemptReauth(reason, options) {
      if (options && options.shouldRefreshAccessToken === false) {
        return _promise2.default.resolve(false);
      }

      if (reason.statusCode === 401) {
        return _promise2.default.resolve(true);
      }

      return _promise2.default.resolve(false);
    }
  }], [{
    key: 'create',

    /**
     * @returns {AuthInterceptor}
     */
    value: function create() {
      return new AuthInterceptor({ webex: this });
    }
  }]);
  return AuthInterceptor;
}(_httpCore.Interceptor); /*!
                           * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
                           */

exports.default = AuthInterceptor;
