// @webex/webex-core/dist/lib/services/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.ServiceUrl = exports.ServiceHost = exports.ServiceState = exports.ServiceRegistry = exports.ServiceCatalog = exports.Services = exports.ServerErrorInterceptor = exports.ServiceInterceptor = exports.constants = undefined;

var _service = require('./interceptors/service');

Object.defineProperty(exports, 'ServiceInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_service).default;
  }
});

var _serverError = require('./interceptors/server-error');

Object.defineProperty(exports, 'ServerErrorInterceptor', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_serverError).default;
  }
});

var _services = require('./services');

Object.defineProperty(exports, 'Services', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_services).default;
  }
});

var _serviceCatalog = require('./service-catalog');

Object.defineProperty(exports, 'ServiceCatalog', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_serviceCatalog).default;
  }
});

var _serviceRegistry = require('./service-registry');

Object.defineProperty(exports, 'ServiceRegistry', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_serviceRegistry).default;
  }
});

var _serviceState = require('./service-state');

Object.defineProperty(exports, 'ServiceState', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_serviceState).default;
  }
});

var _serviceHost = require('./service-host');

Object.defineProperty(exports, 'ServiceHost', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_serviceHost).default;
  }
});

var _serviceUrl = require('./service-url');

Object.defineProperty(exports, 'ServiceUrl', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_serviceUrl).default;
  }
});

var _webexCore = require('../../webex-core');

var _constants = require('./constants');

var constants = _interopRequireWildcard(_constants);

var _services2 = _interopRequireDefault(_services);

var _serverError2 = _interopRequireDefault(_serverError);

var _service2 = _interopRequireDefault(_service);

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

(0, _webexCore.registerInternalPlugin)('services', _services2.default, {
  interceptors: {
    ServiceInterceptor: _service2.default.create,
    ServerErrorInterceptor: _serverError2.default.create
  }
}); /*!
     * Copyright (c) 2015-2019 Cisco Systems, Inc. See LICENSE file.
     */
exports.constants = constants;
