// @webex/webex-core/dist/lib/services/service-state.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _constants = require('./constants');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * The state of a specific catalog to be used by {@link ServiceState}.
 *
 * @typedef {Record<string, boolean>} CatalogState
 * @property {boolean} CatalogState.collecting - If the catalog is collecting.
 * @property {boolean} CatalogState.ready - If the catalog is ready.
 */

/**
 * @class
 * @classdesc - Manages the state of the service catalogs for a webex instance.
 */
var ServiceState = function () {
  /**
   * Generate a new {@link ServiceState}.
   *
   * @public
   * @constructor
   * @memberof ServiceState
   */
  function ServiceState() {
    var _this = this;

    (0, _classCallCheck3.default)(this, ServiceState);

    // Iterate over the possible catalog names and generate their states.
    _constants.SERVICE_CATALOGS.forEach(function (catalog) {
      _this[catalog] = ServiceState.generateCatalogState();
    });
  }

  /**
   * Set a catalog to be collecting or not.
   *
   * @public
   * @memberof ServiceState
   * @param {string} catalog - Catalog to target.
   * @param {boolean} collecting - If the target is collecting or not.
   * @returns {undefined}
   */


  (0, _createClass3.default)(ServiceState, [{
    key: 'setCollecting',
    value: function setCollecting(catalog, collecting) {
      // Validate that the catalog state exists.
      if (this[catalog]) {
        // Set the 'collecting' status of the catalog state.
        this[catalog].collecting = collecting;
      }
    }

    /**
     * Set a catalog to be ready or not.
     *
     * @public
     * @memberof ServiceState
     * @param {string} catalog - Catalog to target.
     * @param {boolean} ready - If the target is ready or not.
     * @returns {undefined}
     */

  }, {
    key: 'setReady',
    value: function setReady(catalog, ready) {
      // Validate that the catalog state exists.
      if (this[catalog]) {
        // Set the 'ready' status of the catalog state.
        this[catalog].ready = ready;
      }
    }

    /**
     * Generate a {@link CatalogState}.
     *
     * @public
     * @static
     * @memberof ServiceState
     * @returns {CatalogState} - The generated {@link CatalogState}.
     */

  }], [{
    key: 'generateCatalogState',
    value: function generateCatalogState() {
      return {
        collecting: false,
        ready: false
      };
    }
  }]);
  return ServiceState;
}();

exports.default = ServiceState;
